File: tools/packages/fetch-client/packages_py/fetch_client/src/fetch_client/core/base_client.py
"""
Core HTTP client implementation based on httpx.
"""
import asyncio
import json
import logging
from contextlib import asynccontextmanager
from typing import Any, AsyncGenerator, Dict, Optional, Union

import httpx

from ..config import ClientConfig, resolve_config, ResolvedConfig
from ..auth.auth_handler import create_auth_handler, AuthHandler
from ..types import FetchResponse, RequestContext, RequestOptions

logger = logging.getLogger(__name__)

# Constants
LOG_PREFIX = "[FetchClient]"

def _format_body(body: Any) -> str:
    """
    Format body for logging safeguards against binary data.
    """
    if body is None:
        return "<empty>"
    if isinstance(body, (bytes, bytearray)):
        return f"<binary data: {len(body)} bytes>"
    if isinstance(body, str):
        try:
            # Try to pretty print if it looks like JSON
            if body.strip().startswith(("{", "[")):
                 return json.dumps(json.loads(body), indent=2)
        except json.JSONDecodeError:
            pass
        # Truncate long strings
        if len(body) > 5000:
             return body[:5000] + "... (truncated)"
        return body
    if isinstance(body, dict):
        return json.dumps(body, indent=2)
    return str(body)

class BaseClient:
    """
    Base HTTP client wrapping httpx.AsyncClient.
    """
    def __init__(self, config: ClientConfig):
        self._config_raw = config
        self._config: ResolvedConfig = resolve_config(config)
        self._client: Optional[httpx.AsyncClient] = config.httpx_client
        self._auth_handler: Optional[AuthHandler] = None
        
        if self._config.auth:
            self._auth_handler = create_auth_handler(self._config.auth)

        # Flag to track if we own the client (created it)
        self._own_client = self._client is None

    async def connect(self) -> None:
        """Initialize the client if needed."""
        if self._client:
            return

        # Create httpx client
        timeout = httpx.Timeout(
            connect=self._config.timeout.connect,
            read=self._config.timeout.read,
            write=self._config.timeout.write,
            pool=self._config.timeout.pool
        )
        
        self._client = httpx.AsyncClient(
            base_url=self._config.base_url,
            timeout=timeout,
            headers=self._config.headers,
            # Force HTTP/1.1 for broad compatibility unless specified otherwise contextually
            # But httpx defaults are usually fine.
            follow_redirects=True, 
        )
    
    async def close(self) -> None:
        """Close the client if we own it."""
        if self._own_client and self._client:
            await self._client.aclose()
            self._client = None

    async def __aenter__(self) -> "BaseClient":
        await self.connect()
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb) -> None:
        await self.close()

    async def request(self, options: RequestOptions) -> FetchResponse:
        """Execute a request."""
        if not self._client:
            await self.connect()
            
        assert self._client is not None
        
        # Prepare URL
        url = options.get("url", "")
        method = options.get("method", "GET")
        
        # Prepare Headers
        headers = options.get("headers", {}).copy()
        
        # Add content type if JSON is present and header missing
        if "json" in options and "Content-Type" not in headers:
            headers["Content-Type"] = "application/json"

        # Prepare Body
        data = options.get("data")
        json_body = options.get("json")
        
        # Resolve Auth
        if self._auth_handler:
            context: RequestContext = {
                "method": method,
                "url": f"{self._config.base_url}/{url}".rstrip("/"), # approx full URL
                "headers": headers,
                "body": json_body if json_body is not None else data
            }
            auth_headers = self._auth_handler.get_header(context)
            if auth_headers:
                headers.update(auth_headers)

        # Logging
        logger.debug(f"{LOG_PREFIX} Request: {method} {url}")
        
        try:
            response = await self._client.request(
                method=method,
                url=url,
                headers=headers,
                params=options.get("params"),
                content=data,
                json=json_body,
            )
            
            # Response handling
            res_data = None
            try:
                res_data = response.json()
            except json.JSONDecodeError:
                res_data = response.text
                
            return FetchResponse(
                status=response.status_code,
                status_text=response.reason_phrase,
                headers=dict(response.headers),
                url=str(response.url),
                data=res_data,
                ok=response.is_success
            )
            
        except httpx.RequestError as e:
            logger.error(f"{LOG_PREFIX} Request failed: {e}")
            raise e

File: packages_mjs/fetch-client/src/core/base-client.ts
/**
 * Core HTTP client implementation based on undici.
 */
import { Client, Dispatcher, Pool, request } from 'undici';
import { AuthHandler, createAuthHandler } from '../auth/auth-handler.js';
import { ClientConfig, ResolvedConfig, resolveConfig } from '../config.js';
import { FetchResponse, RequestContext, RequestOptions } from '../types.js';

const LOG_PREFIX = '[FetchClient]';

/**
 * Format body for logging safeguards against binary data.
 * FIX GAP-2.2: Implement safe body logging parity with Python.
 */
function _formatBody(body: any): string {
    if (body === undefined || body === null) {
        return '<empty>';
    }

    // Binary checks
    if (body instanceof Buffer || body instanceof Uint8Array || body instanceof ArrayBuffer) {
        return `<binary data: ${body.byteLength || (body as Buffer).length} bytes>`;
    }

    // String checks
    if (typeof body === 'string') {
        try {
            if (body.trim().startsWith('{') || body.trim().startsWith('[')) {
                return JSON.stringify(JSON.parse(body), null, 2);
            }
        } catch {
            // Not JSON
        }

        if (body.length > 5000) {
            return body.slice(0, 5000) + '... (truncated)';
        }
        return body;
    }

    // Object/Array
    if (typeof body === 'object') {
        try {
            return JSON.stringify(body, null, 2);
        } catch (e) {
            return '[Circular or Unserializable Object]';
        }
    }

    return String(body);
}

export class BaseClient {
    private config: ResolvedConfig;
    private dispatcher?: Dispatcher;
    private authHandler?: AuthHandler;
    private ownDispatcher: boolean = false;
    private origin: string;
    private basePath: string;

    constructor(config: ClientConfig) {
        this.config = resolveConfig(config);

        // Parse Origin and BasePath to safely support baseUrls with paths (e.g. https://api.com/v1)
        // undici.Pool only accepts origin, so we must manually prepend basePath to requests.
        try {
            const url = new URL(this.config.baseUrl);
            this.origin = url.origin;
            this.basePath = url.pathname;
            // Normalize basePath: remove trailing slash unless it's just root
            if (this.basePath.endsWith('/') && this.basePath.length > 1) {
                this.basePath = this.basePath.slice(0, -1);
            }
            if (this.basePath === '/') {
                this.basePath = '';
            }
        } catch (e) {
            // Fallback (e.g. relative URLs? mostly unlikely in this context)
            this.origin = this.config.baseUrl;
            this.basePath = '';
        }

        // Setup Auth
        if (this.config.auth) {
            this.authHandler = createAuthHandler(this.config.auth);
        }

        // Setup Dispatcher if provided
        // casting because undici types might mismatch strictly with 'any'
        this.dispatcher = config.dispatcher;
        this.ownDispatcher = !this.dispatcher;
    }

    /**
     * Initialize dispatcher if needed.
     */
    private ensureDispatcher() {
        if (this.dispatcher) return;

        // Create default client/pool
        // Undici Client is for single origin, Pool is for single origin? 
        // Agent is for multiple? Undici 'request' uses global dispatcher usually.
        // If baseUrl is set, we can use a Client optimized for that origin.

        // Using a Pool for the base URL is efficient for repeated requests.
        this.dispatcher = new Pool(this.origin, {
            connect: {
                timeout: this.config.timeout.connect
            },
            headersTimeout: this.config.timeout.read,
            bodyTimeout: this.config.timeout.read,
            // Disable keep-alive timeouts or set appropriately?
        });
        this.ownDispatcher = true;
    }

    async close(): Promise<void> {
        if (this.ownDispatcher && this.dispatcher) {
            await this.dispatcher.close();
            this.dispatcher = undefined;
        }
    }

    async request<T = any>(options: RequestOptions): Promise<FetchResponse<T>> {
        const { response, path } = await this.dispatchRequest(options);

        // Response parsing
        let data: any;
        const contentType = response.headers['content-type'] || '';
        const text = await response.body.text();

        try {
            if (contentType.includes('application/json')) {
                // Handle empty body for JSON content type
                data = text ? JSON.parse(text) : {};
            } else {
                data = text;
            }
        } catch {
            data = text;
        }

        return {
            status: response.statusCode,
            statusText: '',
            headers: response.headers as Record<string, string>,
            url: `${this.config.baseUrl}${path}`,
            data,
            ok: response.statusCode >= 200 && response.statusCode < 300
        };
    }

    /**
     * Allows streaming response consumption without buffering the entire body.
     */
    async rawStream(options: RequestOptions): Promise<FetchResponse<ReadableStream>> {
        const { response, path } = await this.dispatchRequest(options);

        return {
            status: response.statusCode,
            statusText: '',
            headers: response.headers as Record<string, string>,
            url: `${this.config.baseUrl}${path}`,
            data: response.body, // Return the raw ReadableStream
            ok: response.statusCode >= 200 && response.statusCode < 300
        };
    }

    /**
     * Internal dispatch helper to reuse logic for request/stream.
     */
    protected async dispatchRequest(options: RequestOptions): Promise<{ response: any, path: string }> {
        this.ensureDispatcher();

        const method = options.method || 'GET';
        // URL handling: passed url is relative path usually, but undici Pool expects path.
        // If baseUrl was used to create Pool, we just pass path.
        const reqPath = options.url || '/';

        // Headers
        const headers: Record<string, string> = {
            ...this.config.headers,
            ...(options.headers || {})
        };

        // Content Type
        if (options.json && !headers['content-type']) {
            headers['content-type'] = 'application/json';
        }

        // Body
        let body: any = options.body;
        if (options.json) {
            body = JSON.stringify(options.json);
        }

        // Auth
        if (this.authHandler) {
            const context: RequestContext = {
                method,
                url: `${this.config.baseUrl}${reqPath.startsWith('/') ? '' : '/'}${reqPath}`,
                headers,
                body: options.json || body
            };
            const authHeaders = this.authHandler.getHeader(context);
            if (authHeaders) {
                Object.assign(headers, authHeaders);
            }
        }

        // Logging
        // console.debug(`${LOG_PREFIX} Request: ${method} ${path}`);

        try {
            // If we are using a Pool (bound to baseUrl), we pass just the path.
            // But if dispatcher is Agent (or MockAgent), we typically need full URL for it to route correctly.
            // Undici MockAgent intercepts based on origin + path.

            // Check for Pool OR MockPool (which behaves like Pool regarding path)
            const isPool = this.dispatcher instanceof Pool || this.dispatcher?.constructor.name === 'MockPool';

            // Calculate effective path including params from baseUrl if any (e.g. /api/v1)
            // If isPool (bound to origin), we must supply full path /api/v1/users
            // If !isPool (Agent), we supply full URL https://host/api/v1/users

            const effectivePath = this.basePath
                ? `${this.basePath}${reqPath.startsWith('/') ? '' : '/'}${reqPath}`
                : reqPath;

            const targetUrl = isPool ? effectivePath : `${this.origin}${effectivePath.startsWith('/') ? '' : '/'}${effectivePath}`;

            // console.log('Requesting:', targetUrl, 'isPool:', isPool);

            const response = await this.dispatcher!.request({
                method,
                path: targetUrl,
                query: options.query,
                headers,
                body,
                headersTimeout: options.timeout || this.config.timeout.read
            });

            return { response, path: reqPath };

        } catch (e) {
            // console.error(`${LOG_PREFIX} Request failed:`, e);
            throw e;
        }
    }
}
